applications/ExaquteSandboxApplication/python_scripts/WindGenerator/onthefly_mpi_kratos_inlet_process.py

```diff
@@ -0,0 +1,88 @@
+
+    def __init__(self, Model, settings):
+        for name, value in settings.items():
+            setattr(self, name, value)
+        #mpi modelpart
+        self.mpi_model_part = Model[self.inlet_model_part_name]
+        self.data_comm = self.mpi_model_part.GetCommunicator().GetDataCommunicator()
+        # serial inlet
+        if not Model.HasModelPart("wind_inlet_serial"):
+            self.model_part = Model.CreateModelPart("wind_inlet_serial")
+            self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VELOCITY)
+            self.model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PARTITION_INDEX)
+            if self.data_comm.Rank() == 0:
+                KratosMultiphysics.ModelPartIO(self.wind_inlet_aux_file_name).ReadModelPart(self.model_part)
+        else:
+            self.model_part = Model.GetModelPart("wind_inlet_serial")
+
+        ParallelFillCommunicator = KratosMPI.ParallelFillCommunicator(self.model_part, self.data_comm)
+        ParallelFillCommunicator.Execute()
+
+        self.x0 = 0.0
+        if self.data_comm.Rank() == 0:
+            if len(self.inlet_nodes) > 0:
+                # define dimensions of the inlet domain
+                y_extent = get_extent(self.inlet_nodes, key=lambda node: node.Y)
+                z_extent = get_extent(self.inlet_nodes, key=lambda node: node.Z)
+                setattr(self, 'ly', y_extent.upper - y_extent.lower)
+                setattr(self, 'lz', z_extent.upper - z_extent.lower)
+                self.mean_profile = self.CreateMeanProfile()
+                # x_extent = Extent(0, self.time_interval_length * self.mean_profile.bulk_wind_speed)
+                # setattr(self, 'lx', x_extent.upper - x_extent.lower)
+                self.inlet_position = 0.0
+
+                # define wind field and map
+                grid_dimensions = [self.lx, self.ly, self.lz]
+                self.wind = GenerateWind(self.friction_velocity, self.reference_height, grid_dimensions, self.wind_grid_levels, self.seed)
+                self.block_num=0
+                self.blend_region, self.mappers = self.Create3DMappers(self.wind)
+
+
+    def UpdateInletPosition(self):
+        dt = self.mpi_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
+        self.inlet_position += dt * self.mean_profile.bulk_wind_speed
+
+        #If the end of the time block is reached, generate a new wind block and reset map
+        if self.inlet_position >= self.x0 + self.lx:
+            self.x0 += self.lx
+            self.blend_region, self.mappers = self.Create3DMappers(self.wind, self.blend_region)
+
+    def ApplyRamp(self):
+        time = self.mpi_model_part.ProcessInfo[KratosMultiphysics.TIME]
+        if time < self.ramp_time:
+            scal = time / self.ramp_time
+            for node in self.inlet_nodes:
+                for var, _ in self.mappers:
+                    vel = node.GetSolutionStepValue(var)
+                    node.SetSolutionStepValue(var, scal * vel)
```

Comment:
Are these two methods required to be defined here? Can't we use the base class methods?

Reply:
We should update the methods to depend on the main model part, note that TIME and DELTA_TIME is read from the `self.mpi_model_part`, which is not the `self.inlet_model_part` that is used to compute the wind. 

So do something like this in the base class:

  ```
  def UpdateInletPosition(self, this_model_part):
        dt = this_model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
        self.inlet_position += dt * self.mean_profile.bulk_wind_speed

        #If the end of the time block is reached, generate a new wind block and reset map
        if self.inlet_position >= self.x0 + self.lx:
            self.x0 += self.lx
            self.blend_region, self.mappers = self.Create3DMappers(self.wind, self.blend_region)
```

Either this or have `dt` as input parameter

Reply:
Ok I did not catch the usage of `mpi_model_part`. To me, it's cleaner to do it as you suggest in this comment. No?